rackcity/views/asset_views.py

```diff
@@ -165,6 +161,30 @@ def asset_add(request):
             status=HTTPStatus.BAD_REQUEST
         )
     rack_id = serializer.validated_data['rack'].id
+    try:
+        rack = Rack.objects.get(id=rack_id)
+    except ObjectDoesNotExist:
+        return JsonResponse(
+            {
+                "failure_message":
+                    Status.MODIFY_ERROR.value +
+                    "Rack" + GenericFailure.DOES_NOT_EXIST.value,
+                "errors": "No existing rack with id="+str(rack_id)
+            },
+            status=HTTPStatus.BAD_REQUEST
+        )
+    if not user_has_asset_permission(request.user, rack.datacenter):
+        return JsonResponse(
+            {
+                "failure_message":
+                    Status.AUTH_ERROR.value + AuthFailure.ASSET.value,
+                "errors":
+                    "User " + request.user.username +
+                    " does not have asset permission in datacenter id="
+                    + str(rack.datacenter.id)
+            },
+            status=HTTPStatus.UNAUTHORIZED
+        )
```

Comment:
I don't love that you can fail in other ways BEFORE failing because you don't have permission but I see why you did it and don't think it's a big deal

Reply:
yeah per-datacenter kind of forces it to be like this, there's no use in checking for global asset permission at the beginning and need to make sure there's a valid rack in the request before you can check if the user has permission in that rack. ceste la vie